### ISSUE TITLE
Huge lag with AutoLink - Give possibility do disable mode with in game settings

### ISSUE THREAD

**Issue by kaostheory78** [Epp-s-Satisfactory-Mods/AutoLink#9]
Hello, 

First of all I want to thank your very much for your contribution with this mod.
I was very disapointed when I realised this feature was missing from the game and I was so glad to see that you made a mod to get over it.

I'm currently using your mod to create a huge road going through all the map.
This road have several different floors, with a floor dedicated to hypertubes (8 of them), and a floor to conveyor belts (also 8 of them).

As my road is growing, as my hypertube length are growing, and I'm experiencing huge latency / freeze for few minutes when I add a piece of blueprints to the existing road, and they're getting bigger as the road is growing.
I noticed the AutoLink mod, is going through the whole length of the hypertube / convoyer belts to correct issues on the whole hypertube.
This is super cool but is creating performance issues on my side (even if I could'nt find a IO device from computer overloading during this process).

As it may seems quite complicated to solve the performance issues, I was wondering if you could add an in game setting to enable and disable the mod (/ mode feature). As the mod will recover the whole section connected to the new blueprint created, I could do the recovery process only once at the end.
May be another additional request could be to add a 'boundary mode' to the Autolink, to just checks the pieces connected to the boudaries of the blueprint instead of the whole hyper tube / pipe / ..

Again I want to thank you because it is already super great, and I would be even more greatful if you can make somehow a small update to help me get over the lags.

Thanks & Regards, 
Quentin

**Comment by Epp-code** [comment]
Hi Quentin - thanks for the feedback.  Which version of the mod are you using and do you have any other mods installed?  Are you sure that the hypertubes have anything to do with the lag?  I'm pretty sure it would be the belts that are causing the issue, as the game chooses to rebuild the entire belt chain for each belt every time you add a new one to it but doesn't do that for the other types of connections, AFAIK.  That said, there could be details of the hypertube system that I don't know about because I didn't encounter them when making the mod.  But I'm also sure there are optimizations that could be made on the mod's end simply as far as how it searches for connections.

Let's focus on your specific case, though. If you're building a blueprint with 8 belts, it theoretically shouldn't lag more than roughly 8x the time it would take to build one of those belts normally.  Do you experience any lag at all when you manually add a single belt to these long blueprint chains?  And is there different amounts of lag depending on whether you manually add to the beginning or end of the belt chain?  What about the lag when you build the blueprint at the beginning or end of the chain?

I have had another player report lag when they add their blueprint to one end of a super-long belt chain but not the other - their workaround was to always build on one end, so that could be something to do in the meantime.

Finally, is there any way you could share your save file and blueprint so I can load it up and debug on my end?  It will be much easier to identify performance issues if I can play around with it.

As far as the mode suggestion, I think that opens up a huge, separate can of worms to deal with.  Let's start with looking for the specific performance issues to determine whether it's fixable or a game limitation.

**Comment by kaostheory78** [comment]
Hello again, 

So I made few trials as you requested.
Actually the idea I was proposing is not working, when building few bleuprints section with the mode disabled and one later on at the end with the mode enabled did not retrieve the connection of the others sections.

I'm not sure the issue is specifically linked to the hypertube, and actually the lagg also appears during the convoyer construction, so it might be both. I would have to modify the blueprint to confirm this.

I tried to prolongate both conveyer belt and hypertube by hand, and it seemd to work fine, but I don't see an orange construction light walk through the whole hypertube and convoyer belts.

I will also try to build at the other end of the road. I haven't tried yet.
But if it is related to conveyor belt direction, I have both in the blueprint...

Here is the save file:
[save file.zip](https://github.com/user-attachments/files/17653483/save.file.zip)

Here are the two blueprints involved:
![image](https://github.com/user-attachments/assets/01cb3cb9-c840-4384-9efa-df6a7e08a88e)

I don't know if you will popup at the right position on the map.
But here is my location : 
[-214 ; -2002]

The road is pretty far up in the sky, don't judge the design, it's my friend I'm playing with who did this whole shit :) 
(And he is arguing that the design is not yet finished)

btw I'm using the version 0.5.3
Please let me know if I can help further.

**Comment by Epp-code** [comment]
Thanks for the save file!  The blueprints are actually stored in separate files on your computer.  They are under ...FactoryGame\Saved\SaveGames\blueprints\<session name>.  Will you please also send me the files for the ones you're using?

**Comment by kaostheory78** [comment]
Oh ok  here are the blueprints:
[blueprints.zip](https://github.com/user-attachments/files/17654080/blueprints.zip)

**Comment by Epp-code** [comment]
Hi there Quentin, I apologize but I haven't gotten to looking at this, yet.  I appreciate the additional testing you've done, as it will help me narrow things down.  I still plan to look but it may be a bit.

**Comment by Epp-code** [comment]
Hi Quentin, I wanted to let you know I have reproduced the issue using the save you provided.  At the moment, I am still totally confused as to why this is happening.  The mod itself does all of its work in the normal time but the blueprint construction animation seems to take a very long time (it looks like the resources are flying to the wrong place?) and THEN the lag kicks in and doesn't go away until I reload the game.  Can you confirm this is the same as what you're seeing?  I will continue to investigate but I'm not yet sure what, if anything, I can do about it.

**Comment by kaostheory78** [comment]
Hello, 

The animation indeed takes a very long time, I think in the save files I have you it takes up to 10 minutes. 
But if you put only 5 blueprints successively it will take less time, and you will see that the animation taking a lot of time is the pipeline animation, the whole section of the pipeline starting in the constructed blueprints up to the end of the complete pipeline will change to orange with the animation going through the whole length. 

I did some more tests on my side. And I find out that i have the same thing on the conveyor belt if I do not have splitter / merger on the line. 
The merger / splitter seems to cut the "going-through" animation. 

I haven't tested it yet but I'm wondering if it could be linked to the facts that in my blueprint the  conveyor poles / hypertube poles overlaps outside the blueprint. I put the poles at the very edge of the blueprint. 
I was thinking maybe by doing that, the mod / the game, would consider the section outside of the blueprint as part of the blueprint and would go through its whole section for the built animation. I haven't tried yet this theory.

**Comment by Epp-code** [comment]
Yeah, I think something about where I'm auto-linking in the blueprint construction process is causing it to think ALL the hypertubes are part of the blueprint that was just built when it calculates the construction animations.  I don't think there's anything wrong with your blueprint - there's probably some cumulative effect I need to uncover.  But I still am at a complete loss for why the performance would tank so hard AFTER all of this is done.  I think I need to do a lot of experimentation with the game and, unfortunately, I now won't have access to my computer until the new year.  I'm sorry this has been such a long process and will continue to be longer - thanks for your patience and experimentation.  I have several ideas about how to investigate and fix this that I plan to test as soon as I can.

**Comment by Epp-code** [comment]
Well, it took quite a while to find the specific game code causing the issue, but it should be fix in v0.5.6, which I just uploaded.  I tested using your save and blueprints - all the materials went to the right place and the huge performance dip didn't happen.  I'm gonna close this ticket - please open a new one if you encounter further issues!

**Comment by kaostheory78** [comment]
Thank you very much ! ❤️
